algopy_testing/src/algopy/primitives/uint64.py

```diff
@@ -0,0 +1,735 @@
+        if other_value == 0:
+            raise UInt64ZeroDivisionError(other_value)
+        self.value %= other_value
+        return self
+
+    def __pow__(self, other: int | UInt64, modulo: int | UInt64 | None = None) -> UInt64:
+        """
+        Raise the UInt64 to the power of another UInt64 or int.
+
+        Args:
+            other (int | UInt64): The exponent to raise the UInt64 to.
+            modulo (int | UInt64 | None, optional): The modulus to apply to the result.
+            Defaults to None.
+
+        Raises:
+            UInt64OverflowError: If the power result exceeds the maximum value for a UInt64.
+
+        Returns:
+            UInt64: The result of the power operation.
+        """
+        exp = self._get_value(other)
+        if modulo is not None:
+            mod = self._get_value(modulo)
+            result = pow(self.value, exp, mod)
+        else:
+            result = pow(self.value, exp)
+        if result > MAX_UINT64:
+            raise UInt64OverflowError(result)
+        return UInt64(result)
+
+    def __rpow__(self, other: int) -> UInt64:
+        """
+        Raise an int to the power of the UInt64.
+
+        Args:
+            other (int): The base to raise to the power of the UInt64.
+
+        Raises:
+            UInt64OverflowError: If the power result exceeds the maximum value for a UInt64.
+
+        Returns:
+            UInt64: The result of the power operation.
+        """
+        result: int = pow(other, self.value)
+        if result > MAX_UINT64:
+            raise UInt64OverflowError(result)
+        return UInt64(result)
+
+    def __ipow__(self, other: int | UInt64, modulo: int | UInt64 | None = None) -> Self:
+        """
+        In-place power of the UInt64 with another UInt64 or int.
+
+        Args:
+            other (int | UInt64): The exponent to raise the UInt64 to.
+            modulo (int | UInt64 | None, optional): The modulus to apply to the result.
+            Defaults to None.
+
+        Raises:
+            OverflowError: If the power result exceeds the maximum value for a UInt64.
+
+        Returns:
+            Self: The result of the in-place power operation.
+        """
+        exp = self._get_value(other)
+        self.value = pow(self.value, exp, MAX_UINT64 + 1)
+        if self.value > MAX_UINT64:
+            raise OverflowError("UInt64 power result exceeds maximum value")
+        return self
+
+    def __lshift__(self, other: int | UInt64) -> UInt64:
+        """
+        Left shift the UInt64 by another UInt64 or int.
+
+        Args:
+            other (int | UInt64): The amount to shift the UInt64 by.
+
+        Raises:
+            ValueError: If the shift amount is negative.
+
+        Returns:
+            UInt64: The result of the left shift operation.
+        """
+        shift = self._get_value(other)
+        if shift < 0:
+            raise UInt64BitShiftInvalidValueError("Shift amount must be non-negative")
```

Comment:
this doesn't need a specific error, just construct a UInt64 from the right value, which is underflow.
negative constants won't compile anyway, and we don't need to guarantee any semantic compatibility if the code doesn't compile

Reply:
Have opted for using builtin python errors for now, we can always move towards more specific types if required